docs/build_file.md

```diff
@@ -0,0 +1,36 @@
+# build.tl file
+Teal has the ability to run a custom function at compile time. This file by default is called `build.tl` but the name can be changed by setting `build_file` in `tlconfig.lua`.
+
+This file can be used to automatically generate code/types without the need for external build tools like `make`. 
+
+## Layout
+
+A build.tl file needs at least the following layout:
+```lua
+return {
+    gen_code = function(path:string)
+
+    end
+}
+```
+`gen_code` is the function that will get executed and `path` is the base path where it should store generated teal files.
+
+More keys are planned in the future, which is why the file returns a table rather than it being executed directly.
+
+## Output location
+
+The teal files get stored in a unique folder in `/tmp`, this folder will get cleaned up after the compiling succeeded. The generated teal files will get compiled to lua as normal and will be part of the build output.
+
+You can configure where the lua files will be saved by setting `build_file_output_dir` in `tlconfig.lua`. This uses the directory set by `build_dir` as a base. The default value is `generated_code`.
+
+## Use case
+
+As mentioned earlier, this file can be used to generate types without the need for `make` or other external build tools. A reason why you might is if your teal code consumes an API that has schemas available.
+
+Then you could simply add these schemas to your repo and have the `build.tl` file create types based on these schemas for you. That way you only need to grab a new version of the schemas if they change.
+
+Another use case could be when using a teal version of a library like [pgtyped](https://github.com/adelsz/pgtyped) where you normally need to run a command manually to generate the types and code. Now you can just stick that in the `build.tl` file and forget about it.
+
```

Comment:
I think a small example that actually does something would be helpful here

Reply:
so, just have it create a simple teal file that returns a string or something?